contents/docs/plugins/build/overview.md

```diff
@@ -0,0 +1,73 @@
+---
+title: Overview
+sidebar: Docs
+showTitle: true
+---
+
+A great way to understand what you can do with plugins is to understand how data flows throw plugins.
+
+There's 2 big concepts to remember:
+
+1. Every plugin acts on a single event coming in.
+
+2. Plugins act on events before they are stored.
```

Comment:
There's just one problem with this line in that it's not technically true. I'm not sure if this is written like this for simplicity or just in error. However some plugins act on events before they are stored. Some plugins act on events *at the same time* that the event is being stored.

Basically:

```js
event = await getNextEvent()
event = await processEvent1(event)
event = await processEvent2(event)
await Promise.all([
    storeEventInDatabase(event),
    onEvent1(event),
    onEvent2(event),
])
```

This means also that the figure for the GeoIP plugin example is correct, but the one for the S3 plugin is not. It goes into S3 and our db at the same time, two parallel arrows.

Reply:
Indeed, that's right! I wanted to emphasise how some special functions mutate and others don't.

*I think* it's a technical detail that doesn't add extra information from the user's perspective? (i.e., I don't think knowing this will help them make better plugins? / understand plugins?)

I could do a footnote saying how this isn't exactly true?

Reply:
I guess a footnote would suffice. The only technically important thing to clarify here is that even if an `onEvent` fails, the event still ends up in the database (as they are running in parallel). Otherwise people might get the wrong idea, and believe that a failure here will block the event.

Reply:
Hmm, makes sense, that's the crux: there's no data loss. I'll highlight that here, and skip the footnote about the extra technical details.

Reply:
Actually, we might need to explain this quite clearly for everything to click in place.

Inside events we have three special properties: `$plugins_deferred`, `$plugins_failed`, `$plugins_succeeded`:

![image](https://user-images.githubusercontent.com/53387/122044307-a59d9800-cddc-11eb-9299-2b8d74c7ffd7.png)

The "deferred" literally implies that these plugins were run after (ingestion).

Reply:
I've made a note to add this bit in the developer reference, but not in the high level overview.